src/tempo.py

```diff
@@ -212,42 +197,34 @@ def generate_config(
                 # try with fqdn?
                 "tls_server_name": self._external_hostname,
             }
-            config["ingester_client"] = {"grpc_client_config": tls_config}
-            config["metrics_generator_client"] = {"grpc_client_config": tls_config}
-
-            config["querier"]["frontend_worker"].update({"grpc_client_config": tls_config})
-
-            # this is not an error.
-            config["memberlist"].update(tls_config)
+            config.ingester_client = Client(grpc_client_config=ClientTLSConfig(**tls_config))
+            config.metrics_generator_client = Client(
+                grpc_client_config=ClientTLSConfig(**tls_config)
+            )
+            config.querier.frontend_worker.grpc_client_config = ClientTLSConfig(**tls_config)
+            config.memberlist = config.memberlist.copy(update=tls_config)
 
         return config
 
     def _build_storage_config(self, s3_config: dict):
-        storage_config = {
-            "wal": {
-                # where to store the wal locally
-                "path": self.wal_path
-            },
-            "pool": {
+        storage_config = TraceStorageConfig(
+            # where to store the wal locally
+            wal=WalConfig(path=self.wal_path),
+            pool=PoolConfig(
                 # number of traces per index record
-                "max_workers": 400,
-                "queue_depth": 20000,
-            },
-            "backend": "s3",
-            "s3": {
-                "bucket": s3_config["bucket"],
-                "access_key": s3_config["access-key"],
-                # remove scheme to avoid "Endpoint url cannot have fully qualified paths." on Tempo startup
-                "endpoint": re.sub(
-                    rf"^{urlparse(s3_config['endpoint']).scheme}://",
-                    "",
-                    s3_config["endpoint"],
-                ),
-                "secret_key": s3_config["secret-key"],
-                "insecure": (False if s3_config["endpoint"].startswith("https://") else True),
-            },
-        }
-        return {"trace": storage_config}
+                max_workers=400,
+                queue_depth=20000,
+            ),
+            backend="s3",
+            s3=S3Config(
```

Comment:
I don't think it would make much of a difference if you're referring to the same level of validation we currently do (keys exist with correct value types). 
once we get the `credentials_changed` event, we generate the config (`_update_tempo_cluster`). 
Also, I think it'd be cleaner to keep construction of config with the validation it offers in one place.

Reply:
I do like the idea to keep the validation in one place